sdk/go/common/workspace/templates_ai.go

```diff
@@ -0,0 +1,148 @@
+}
+
+// Represents the response from the Pulumi AI service to retrieve a project template.
+type AIProjectPackageResponse struct {
+	Manifest []string        `json:"manifest"`
+	Files    []AIFilePayload `json:"files"`
+}
+
+// The shape of an individual file in the response from the Pulumi AI service to retrieve a project template.
+type AIFilePayload struct {
+	Name    string `json:"name"`
+	Content string `json:"content"`
+}
+
+func isAITemplateURL(templateNamePathOrURL string) bool {
+	return strings.HasPrefix(templateNamePathOrURL, "ai")
+}
+
+func retrieveAITemplates(templateURL string) (TemplateRepository, error) {
+	var err error
+	// Create a temp dir.
+	var temp string
+	if temp, err = os.MkdirTemp("", "pulumi-template-"); err != nil {
+		return TemplateRepository{}, err
+	}
+
+	parsedURL, err := url.Parse(templateURL)
+	if err != nil {
+		return TemplateRepository{}, err
+	}
+
+	var fullPath string
+	if fullPath, err = RetrieveAITemplateFolder(parsedURL, temp); err != nil {
+		return TemplateRepository{}, fmt.Errorf("failed to retrieve ai folder: %w", err)
+	}
+
+	return TemplateRepository{
+		Root:         temp,
+		SubDirectory: fullPath,
+		ShouldDelete: true,
+	}, nil
+}
+
+func RetrieveAITemplateFolder(templateURL *url.URL, tempDir string) (string, error) {
+	var projectPackageResponse AIProjectPackageResponse
+	serviceURL, conversationID := mapAITemplateURLToServiceURL(templateURL)
+	fetchNoticeLength, _ := fmt.Printf("Fetching project template for conversation %s from Pulumi AI...", conversationID)
+	fullPackageURL := fmt.Sprintf("%s/project", serviceURL)
+	packageRequestBody := AIProjectRequestBody{
+		ConversationID: conversationID,
+		Model:          "gpt-3.5-turbo",
+	}
+	marshalledDependencyBody, err := json.Marshal(packageRequestBody)
+	if err != nil {
+		return "", err
+	}
+	packageRequest, err := http.NewRequest(http.MethodPut, fullPackageURL, bytes.NewReader(marshalledDependencyBody))
+	if err != nil {
+		return "", err
+	}
+	packageResponse, err := http.DefaultClient.Do(packageRequest)
+	if err != nil {
+		return "", err
+	}
+	fmt.Print(strings.Join([]string{"\r", strings.Repeat(" ", fetchNoticeLength), "\r"}, ""))
+	packageResponseBody, err := io.ReadAll(packageResponse.Body)
+	if err != nil {
+		return "", err
+	}
+	err = json.Unmarshal(packageResponseBody, &projectPackageResponse)
+	if err != nil {
+		return "", err
+	}
+	err = writePulumiAITemplate(projectPackageResponse, tempDir)
+	if err != nil {
+		return "", err
+	}
+	return tempDir, nil
+}
+
+func writePulumiAITemplate(packageResponse AIProjectPackageResponse, tempDir string) error {
+	manifestMap := make(map[string]bool)
+	for _, manifestFile := range packageResponse.Manifest {
+		manifestMap[manifestFile] = true
+	}
+	for _, file := range packageResponse.Files {
+		if _, ok := manifestMap[file.Name]; !ok {
+			fmt.Print(fmt.Errorf("file %s not found in manifest", file.Name))
+			continue // skip files not in the manifest, and alert user that they were skipped
```

Comment:
Why is there even a manifest? Just for uniqueness? We can check that here without needing a manifest from the service.

Reply:
My thought was to ensure totality and to provide a contract from the service about what files were available. Happy to ignore it though.

Reply:
Nothing is printed even when `-v 10` is passed while calling `logging.V(1).Infof`